### PR TITLE
replace unsafe use of dict in Multithreaded code.

### PR DESCRIPTION
**Pull request by OkonSamuel**
Closes issue #300 .
### Further Additions in this PR include
1. `CPUThreads` now accepts an (Optional) integer (e.g `CPUThreads(5)` ) which can be used to control the number of threads that are spawned (as spawning threads has some overhead). Most times the default parameters do the trick.
2. A `Chunks` iterator is implemented to help (1) above.
3. The return values of `_evaluate` method now has the right type (You could check this if you like)

## A slight Note on the use of theses Computational Resources in the `evaluate` method
1. `CPU1()` can be used in all cases
2. when using `CPUThreads(n) or CPUThreads()` speed up is only present if the data is small and model involves heavy computations. The package implementing the code for the model must be thread safe. The magnitude of the speedup depends on the number of threads.
3. when using `CPUProcesses` speed up is only present if the data is small and model involves heavy computations. 
   .

### CODE REVIEW

File: src/resampling.jl [hunk 1]
```diff
@@ -609,47 +608,43 @@ end
...
-
+          end
+      end
+        
     end
-  
-    return reduce(vcat, results)
+    results = reduce(vcat, ret) ## has the right Type
+     return results
 end
```
Comment (ablaom) [inline src/resampling.jl:648]: Not sure I understand how this small change effects the return type but I'll take your word for it.

File: src/resampling.jl [hunk 2]
```diff
@@ -701,7 +696,17 @@ function evaluate!(mach::Machine, resampling, weights,
...
         end
     end
-    if acceleration isa CPUThreads
+     if acceleration isa CPUThreads
+        if acceleration.settings === nothing 
+            nthreads = Threads.nthreads()
+            #Tasks have some ovehead so tried to avoid excessive creation
+             # set default threshold at 50
+            n = div(nfolds, 50)
+            # The parameter passes as input to CPUThreads is the ntasks
```
Comment (ablaom) [inline src/resampling.jl:705]: I suggest we make this "50" a global constant declared in src/MLJBase.jl.

File: src/resampling.jl (hunk 1 above)
Comment (OkonSamuel) [reply src/resampling.jl:648]:
```julia
julia> a = Vector(undef,3)
3-element Array{Any,1}:
 #undef
 #undef
 #undef

julia> for i in 1:length(a)
       a[i] = [1,2,3]
       end

julia> a
3-element Array{Any,1}:
 [1, 2, 3]
 [1, 2, 3]
 [1, 2, 3]

julia> reduce(vcat, a)
9-element Array{Int64,1}:
 1
 2
 3
 1
 2
 3
 1
 2
 3
```

Comment (ablaom) [reply src/resampling.jl:648]: Got it, thanks.

Comment (ablaom) [reply src/resampling.jl:648]: Still, I think the compiler still has to box things, no? Look this example:

```julia
function foo(I)
    a = Array{Any}(undef, 2)
    a[1] = I
    a[2] = I
    return [a...]
end

julia> @code_warntype foo(3)
```
```
Variables
  #self#::Core.Compiler.Const(foo, false)
  I::Int64
  a::Array{Any,1}

Body::Any
1 ─ %1 = Core.apply_type(Main.Array, Main.Any)::Core.Compiler.Const(Array{Any,N} where N, false)
│        (a = (%1)(Main.undef, 2))
│        Base.setindex!(a, I, 1)
│        Base.setindex!(a, I, 2)
│   %5 = Core._apply(Base.vect, a)::Any
└──      return %5
```

(not that I think this matters for this PR)

Comment (OkonSamuel) [reply src/resampling.jl:648]: yeah it still boxes things but the `reduce` methods returns the right type reducing performance losses

**Comment by ablaom** [comment]
Okay, sorry for being a little dense, but why do we set the threshold so high? 

So, a common use case would be CV with say, `nfolds` between about 2 and 10; no repeats.  If `RESAMPLING_THRESHOLD=50`, then we get `n=0` every time and so the default that the code picks out is `CPUThreads(1)`, no matter how many threads are available. What am I missing?

**Comment by OkonSamuel** [comment]
> Okay, sorry for being a little dense, but why do we set the threshold so high?
> 
> So, a common use case would be CV with say, `nfolds` between about 2 and 10; no repeats. If `RESAMPLING_THRESHOLD=50`, then we get `n=0` every time and so the default that the code picks out is `CPUThreads(1)`, no matter how many threads are available. What am I missing?

Thats the idea. I set the minimum work load per task is 50.(The idea is that We don't want to use spawn on tasks that finish in less than a second). This choice is based on the assumption that it takes roughly 0.1 sec to call fit per fold.(0.1*50 is well above a second).

**Comment by ablaom** [comment]
But `fit` might take 2 hours or 2 days, right? If `fit` is taking a few seconds or less, we probably don't care about acceleration at all. The exception might be a very small data set and a model with a big hyper-parameter space we want to explore in tuning. But I would say that's an exceptional use-case, not the norm. 

Indeed, on reflection I would be happy to use all available threads when I don't specify the number explicitly (the existing behaviour, I believe)

 [Side note: In any case, we should make sure our testing here is actually using more than one thread]

**Comment by OkonSamuel** [comment]
> But `fit` might take 2 hours or 2 days

Oh that's very true. I guess I'll just set the default to the number of threads.

File: test/utilities.jl [hunk 3]
```diff
@@ -98,5 +98,17 @@ end
...
+    #test for cases with exactly same work as threads
+    @test length(MLJBase.chunks(1:nthreads, nthreads)) == nthreads
+    #test for cases with more work than threads
+    @test length(MLJBase.chunks(1:nthreads + rand(1:nthreads), nthreads)) == nthreads
+    #test for cases with less work than threads
+    @test length(MLJBase.chunks(1:nthreads-1, nthreads)) == nthreads - 1
+end
+end
+
 end # module
```
Comment (ablaom) [inline test/utilities.jl:113]: Cool.

**Pull request merged**